extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -482,11 +485,110 @@ class LifecycleTask extends BackbeatTask {
+                const transition = this._getApplicableTransition({
+                    transitions: rule.Transitions,
+                    lastModified: metadata.LastModified,
+                    currentDate,
+                });
+                store.Transition = this._getStoreTransition({
+                    transition1: transition,
+                    transition2: store.Transition,
+                    lastModified: metadata.LastModified,
+                });
+            }
+            // TODO: Add support for NoncurrentVersionTransitions.
             return store;
         }, {});
+        // Do not transition to a location where the object is already stored.
+        if (applicableRules.Transition &&
+            applicableRules.Transition.StorageClass === metadata.StorageClass) {
+            applicableRules.Transition = undefined;
+        }
+        return applicableRules;
         /* eslint-enable no-param-reassign */
     }
 
+    /**
+     * Get the Unix timestamp of the given date.
+     * @param {string} date - The date string to convert to a Unix timestamp
+     * @return {number} - The Unix timestamp
+     */
+    _getTimestamp(date) {
+        return new Date(date).getTime();
+    }
+
+    /**
+     * Find the Unix time at which the transition should occur.
+     * @param {object} transition - A transition from the lifecycle transitions
+     * @param {string} lastModified - The object's last modified date
+     * @return {number|undefined} - The normalized transition timestamp
+     */
+    _getTransitionTimestamp(transition, lastModified) {
+        if (transition.Date !== undefined) {
+            return this._getTimestamp(transition.Date);
+        }
+        if (transition.Days !== undefined) {
+            const lastModifiedTime = this._getTimestamp(lastModified);
+            const oneDay = 24 * 60 * 60 * 1000; // Milliseconds in a day.
+            return lastModifiedTime + (transition.Days * oneDay);
+        }
+        return undefined;
+    }
+
+    /**
+     * Find the most relevant trantition rule for the given transitions array.
+     * @param {object} params - The function parameters
+     * @param {array} params.transitions - Array of lifecycle rule transitions
+     * @param {string} params.lastModified - The object's last modified
+     * date
+     * @return {object} The most applicable transition rule
+     */
+    _getApplicableTransition(params) {
+        const { transitions, lastModified, currentDate } = params;
+        return transitions.reduce((result, transition) => {
+            const isApplicable = // Is the transition time in the past?
+                this._getTimestamp(currentDate) >=
+                this._getTransitionTimestamp(transition, lastModified);
+            if (!isApplicable) {
+                return result;
+            }
+            if (!result) {
+                return transition;
+            }
+            const delta =
```

Comment:
Couldn't we rely on `_getStoreTransition` to do the recency check? It would avoid comparing dates at two different places for the same purpose.

Also instead of using `delta` variable you can simply compare timestamps directly.

Reply:
Good call, I'll make that a more generic comparison function.